Detectors/MUON/MCH/Tracking/src/TrackFinderOriginal.cxx

```diff
@@ -0,0 +1,1346 @@
+    if (removeTrack) {
+      print("Removing candidate at position #", getTrackIndex(itTrack));
+      itTrack = mTracks.erase(itTrack);
+    } else {
+      print("Candidate at position #", getTrackIndex(itTrack), " is improved");
+      ++itTrack;
+    }
+  }
+}
+
+//_________________________________________________________________________________________________
+void TrackFinderOriginal::finalize()
+{
+  /// Copy the smoothed parameters and covariances into the regular ones
+
+  print("Finalize tracks");
+
+  // The smoother must be enabled to compute the final parameters at each cluster
+  if (!mTrackFitter.isSmootherEnabled()) {
+    LOG(ERROR) << "Smoother disabled --> tracks cannot be finalized";
+    return;
+  }
+
+  for (auto& track : mTracks) {
+    for (auto& param : track) {
+      param.setParameters(param.getSmoothParameters());
+      param.setCovariances(param.getSmoothCovariances());
+    }
+  }
+}
+
+//_________________________________________________________________________________________________
+uint8_t TrackFinderOriginal::requestedStationMask() const
+{
+  /// Get the mask of the requested station, i.e. an integer where
+  /// bit n is set to one if the station n was requested
+
+  uint8_t mask(0);
+
+  for (int i = 0; i < 5; ++i) {
+    if (SRequestStation[i]) {
+      mask |= (1 << i);
+    }
+  }
+
+  return mask;
+}
+
+//_________________________________________________________________________________________________
+int TrackFinderOriginal::getTrackIndex(const std::list<Track>::iterator& itCurrentTrack) const
+{
+  /// return the index of the track pointed to by the given iterator in the list of tracks
+  /// return -1 if it points to mTracks.end()
+  /// return -2 if it points to nothing in the list
+  /// return -3 if the debug level is < 1 as this function is supposed to be used for debug only
+
+  if (mDebugLevel < 1) {
+    return -3;
+  }
+
+  if (itCurrentTrack == mTracks.end()) {
+    return -1;
+  }
+
+  int index(0);
+  for (auto itTrack = mTracks.begin(); itTrack != itCurrentTrack; ++itTrack) {
+    if (itTrack == mTracks.end()) {
+      return -2;
+    }
+    ++index;
+  }
+
+  return index;
+}
+
+//_________________________________________________________________________________________________
+void TrackFinderOriginal::printTracks() const
+{
+  /// print all the tracks currently in the list if the debug level is > 1
+  if (mDebugLevel > 1) {
+    for (const auto& track : mTracks) {
+      track.print();
+    }
+  }
+}
+
+//_________________________________________________________________________________________________
+void TrackFinderOriginal::printTrackParam(const TrackParam& trackParam) const
+{
+  /// print the track parameters if the debug level is > 1
+  if (mDebugLevel > 1) {
+    trackParam.print();
+  }
+}
+
+//_________________________________________________________________________________________________
+template <class... Args>
+void TrackFinderOriginal::print(Args... args) const
```

Comment:
Why not to use FairLogger ?

Reply:
Hi Ruben,

The idea was to have a first working version of the tracking, basically taken from AliRoot (in which TMatrixD were used), that I will now try to improve. Using SMatrix might help indeed, thanks a lot for the suggestion!

About the print method, I wanted a local debugging mode not to pollute (or be polluted by) general debug messages, and without having "if" statements all over the place in my code, so I found this simple solution. But if there is a more generic way to do it in O2 I will move to it for sure. Anyway this "brute force" debugging is useful for development but it does not meant to stay like this.

Cheers,
Philippe